cmd/pow/cmd/ffs_stage.go

```diff
@@ -31,18 +39,94 @@ var ffsStageCmd = &cobra.Command{
 		defer cancel()
 
 		if len(args) != 1 {
-			Fatal(errors.New("you must provide a file path"))
+			Fatal(errors.New("you must provide a file/folder path"))
 		}
 
-		f, err := os.Open(args[0])
-		checkErr(err)
-		defer func() { checkErr(f.Close()) }()
+		fi, err := os.Stat(args[0])
+		if os.IsNotExist(err) {
+			Fatal(errors.New("file/folder doesn't exist"))
+		}
+		if err != nil {
+			Fatal(fmt.Errorf("getting file/folder information: %s", err))
+		}
+		var cid cid.Cid
+		if fi.IsDir() {
+			token := viper.GetString("token")
+			cid, err = addFolder(args[0], token)
+			if err != nil {
+				Fatal(fmt.Errorf("staging folder: %s", err))
+			}
+		} else {
+			f, err := os.Open(args[0])
+			checkErr(err)
+			defer func() { checkErr(f.Close()) }()
 
-		s := spin.New("%s Caching specified file in FFS hot storage...")
-		s.Start()
-		cid, err := fcClient.FFS.Stage(authCtx(ctx), f)
-		s.Stop()
-		checkErr(err)
-		Success("Cached file in FFS hot storage with cid: %s", util.CidToString(*cid))
+			s := spin.New("%s Staging specified file in FFS hot storage...")
+			s.Start()
+			ptrCid, err := fcClient.FFS.Stage(authCtx(ctx), f)
+			s.Stop()
+			checkErr(err)
+			cid = *ptrCid
+		}
+		Success("Staged asset in FFS hot storage with cid: %s", util.CidToString(cid))
 	},
 }
+
+func addFolder(folderPath string, ffsToken string) (cid.Cid, error) {
+	ipfsRevProxy := viper.GetString("ipfsrevproxy")
+	ipfs, err := newDecoratedIPFSAPI(ipfsRevProxy, ffsToken)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("creating IPFS HTTP client: %s", err)
+	}
+
+	stat, err := os.Lstat(folderPath)
+	if err != nil {
+		return cid.Undef, err
+	}
+	ff, err := files.NewSerialFile(folderPath, false, stat)
+	if err != nil {
+		return cid.Undef, err
+	}
+	defer func() {
+		if err := ff.Close(); err != nil {
+			Fatal(fmt.Errorf("closing folder: %s", err))
+		}
+	}()
+	opts := []options.UnixfsAddOption{
+		options.Unixfs.CidVersion(1),
+		options.Unixfs.Pin(false),
+	}
+	pth, err := ipfs.Unixfs().Add(context.Background(), files.ToDir(ff), opts...)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	return pth.Cid(), nil
+}
+
+func newDecoratedIPFSAPI(ipfsRevProxyMaddr, ffsToken string) (*httpapi.HttpApi, error) {
```

Comment:
This returns an IPFS HTTP api, but with a custom client which adds the auth header on every HTTP call.
This header will be checked by the server to allow the request to pass through the IPFS node. If that isn't the case, the request will be Unauthorized (401).